test-data/unit/check-overloading.test

```diff
@@ -1136,3 +1136,24 @@ def f(a: Callable[[], int]) -> None: pass
 @overload
 def f(a: str) -> None: pass
 f(0)  # E: No overload variant of "f" matches argument types [builtins.int]
+
+[case testCustomRedefinitionDecorator]
+from typing import Any, Callable, Type
+
+class Chain(object):
+    def chain(self, function: Callable[[Any], int]) -> 'Chain':
+        return self
+
+class Test(object):
+    do_chain = Chain()
+
+    @do_chain.chain
+    def do_chain(self) -> int:
+        return 2
+
+    @do_chain.chain  # E: Name 'do_chain' already defined
```

Comment:
One possible option is to allow redefinition of all things named ``_`` (there is already two issues and a PR for this) redefinitions of all other names should be an error. Another option is to allow redefinition of all names that were not given an explicit annotation.

In this particular situation I think we should allow this in any case, since "equivalent semantics" is:
```python
def __(self) -> int:
    return 3
do_cain = do_chain.chain(__)
```